### PR TITLE
error handling

### PR DESCRIPTION
**Pull request by nitzantomer**
all errors are grouped into one module, and the error middleware knows how to serialize them

### CODE REVIEW

**Comment by doodyparizada** [comment]
you have an error on travis: ```
scripts/src/mock_client.ts(9,10): error TS2305: Module '"/home/travis/build/kinecosystem/marketplace-server/scripts/src/public/middleware"' has no exported member 'ApiError'.
```

**Comment by nitzantomer** [comment]
@doodyparizada yeah yeah

File: scripts/src/public/services/orders.ts [hunk 1]
```diff
@@ -215,7 +224,7 @@ export async function getOrderHistory(
 
 function openOrderDbToApi(order: db.Order): OpenOrder {
 	if (order.status !== "opened") {
-		throw new Error("only opened orders should be returned");
+		throw OpenedOrdersOnly();
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: Maybe this should be some assert? it's an internal error and a programming bug

File: scripts/src/public/services/orders.ts [hunk 2]
```diff
@@ -231,7 +240,7 @@ function openOrderDbToApi(order: db.Order): OpenOrder {
 
 function orderDbToApi(order: db.Order): Order {
 	if (order.status === "opened") {
-		throw new Error("opened orders should not be returned");
+		throw OpenedOrdersUnreturnable();
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: same here: Maybe this should be some assert? it's an internal error and a programming bug

File: scripts/src/errors.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
+export type ApiError = {
+	code: number;
+	error: string;
+	message: string;
+};
+
+/**
+ * Code additions (/postfix) to be added to the http status code per error.
```
Comment (liorama) [inline scripts/src/errors.ts]: Why the `/` before postfix? (also, although postfix can be used I think the more correct term is suffix...but I'm just being an Ory now...leave it or change it IDC)

File: scripts/src/errors.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+function ConflictError(index: number, message: string) {
+	return new MarketplaceError(409, index, "Conflict", message);
+}
+
+export function ExternalOrderExhausted() {
+	return ConflictError(CODES.Conflict.ExternalOrderExhausted, "User already completed offer, or has a pending order");
+}
+
+function LockedError(index: number, message: string) {
+	return new MarketplaceError(423, index, "Resource Locked", message);
```
Comment (liorama) [inline scripts/src/errors.ts]: Didn't we agree on a 404?

Comment (doodyparizada) [reply scripts/src/errors.ts]: yes

**Pull request merged**